### DISCUSSION TITLE
Implementing Keycloak SSO Integration with Windows NTLM Authentication

### DISCUSSION THREAD

**Discussion by bleriot14** [keycloak/keycloak#33226] [Q&A]
Hello Everyone.

We use Windows systems within our company. I've successfully integrated Keycloak with LDAP in the background.

 I'm able to fetch users from Windows DC servers without any issues. What I want to achieve is to allow users to authenticate using the silent NTLM mechanism, enabling SSO logic without any form. 

Please note that I'm specifically referring to NTLM, not Kerberos. Has anyone among you been able to accomplish this?

**Comment by bleriot14** [comment]
Anyone ?

**Comment by Xulunix** [comment]
I don't think NTLM Authentication is implemented in Keycloak.
NTML is deprecated, insecure and not recommended for use by Microsoft, so i doubt there will be an official implementation in Keycloak at any time in the future.

**Comment by cluck** [comment]
The NTLMv2 protocol is disappearing "on the wire", but NTLMv2 hashes will likely stick around much longer "as a concept" embedded into other protocols.

A new Kerberos concept called "LocalKDC" should eventually help in replacing the NTLMv2 protocol with the Kerberos protocol on the wire, but this is still experimental. Microsoft started shipping a _Kerberos Local Key Distribution Center (LocalKDC)_ Service to Windows Server 2025 but it isn't GA yet and [causing People trouble](https://learn.microsoft.com/en-us/answers/questions/2136070/windows-server-2025-kerberos-local-key-distributio). Meanwhile, LocalKDC and Samba was a hot topic at FOSDEM2025 ([A local authentication hub](https://fosdem.org/2025/events/attachments/fosdem-2025-5618-localkdc-a-general-local-authentication-hub/slides/238662/2025-fosd_md0SPLI.pdf)).

E.g. "Password Hash Synchronisation" with Entra ultimately relies upon NTLMv2 hashes but they're trasferred over a TLS connection where the client first authenticates the server, and the server PBKDF-hashes the NTLMv2 hashes at rest.

From a Keycloak perspective (potentially bridging network-level SSO and web SSO) the concept of an NTLMv2 hash may remain relevant as a concept.

Other frequently cited resources:
https://syfuhs.net/deprecating-ntlm-is-easy-and-other-lies-we-tell-ourselves
https://techcommunity.microsoft.com/blog/windows-itpro-blog/the-evolution-of-windows-authentication/3926848
